{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Postprocessing data from the reanalysis of the COVID-19 multiomics project\n",
    "\n",
    "As part of the case study associated with the data found in https://dx.doi.org/10.1101%2F2020.07.17.20156513, we compare the features discovered by the authors and us.\n",
    "\n",
    "## Summary of our analysis\n",
    "\n",
    "With our pipeline, we analysed the omics data from the original publication, comparing the severity of COVID-19 states (less severe vs more severe) according to the previously described `HFD-45` metric by the authors. We obtained informative features both within and across the omics data. Then, we compare the features identified from our single-omics analyses, multi-omics analyses, and the authors results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup all file paths\n",
    "data_dir = \"../../data/MSV000085703/\"\n",
    "results_dir = \"../../results/MSV000085703/reformatted_output/\"\n",
    "\n",
    "features_path = \"/\".join([\"../../results/MSV000085703/supp_info/\", \"informative_features.txt\"])\n",
    "unknowns_path = \"/\".join([\"../../results/MSV000085703/supp_info/\", \"putative_unknowns.txt\"])\n",
    "feature_importance_all_path = \"/\".join([\"../../results/MSV000085703/supp_info/\", \"feature_importance_all.txt\"])\n",
    "\n",
    "l_fmap_path = \"/\".join([data_dir, \"lipidomics_featuremap.tsv\"]) \n",
    "md_fmap_path = \"/\".join([data_dir, \"metabolomicsdiscovery_featuremap.tsv\"]) \n",
    "mt_fmap_path = \"/\".join([data_dir, \"metabolomicstargeted_featuremap.tsv\"]) \n",
    "p_fmap_path = \"/\".join([data_dir, \"proteomics_featuremap.tsv\"]) \n",
    "t_fmap_path = \"/\".join([data_dir, \"transcriptomics_featuremap.tsv\"]) \n",
    "\n",
    "l_splsda_path = \"/\".join([results_dir, \"lipidome_sPLSDA_max.txt\"])\n",
    "m_splsda_path = \"/\".join([results_dir, \"metabolome_sPLSDA_max.txt\"])\n",
    "p_splsda_path = \"/\".join([results_dir, \"proteome_sPLSDA_max.txt\"])\n",
    "t_splsda_path = \"/\".join([results_dir, \"transcriptome_sPLSDA_max.txt\"])\n",
    "\n",
    "l_diablo_path = \"/\".join([results_dir, \"lipidome_DIABLO_max.txt\"])\n",
    "m_diablo_path = \"/\".join([results_dir, \"metabolome_DIABLO_max.txt\"])\n",
    "p_diablo_path = \"/\".join([results_dir, \"proteome_DIABLO_max.txt\"])\n",
    "t_diablo_path = \"/\".join([results_dir, \"transcriptome_DIABLO_max.txt\"])\n",
    "\n",
    "diablo_path = \"/\".join([results_dir, \"DIABLO_var_keepx_correlations.txt\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the single omics data\n",
    "l_splsda = pd.read_csv(l_splsda_path, sep=\"\\t\")\n",
    "m_splsda = pd.read_csv(m_splsda_path, sep=\"\\t\")\n",
    "p_splsda = pd.read_csv(p_splsda_path, sep=\"\\t\")\n",
    "t_splsda = pd.read_csv(t_splsda_path, sep=\"\\t\")\n",
    "\n",
    "# load the multi omics data\n",
    "l_diablo = pd.read_csv(l_diablo_path, sep=\"\\t\")\n",
    "m_diablo = pd.read_csv(m_diablo_path, sep=\"\\t\")\n",
    "p_diablo = pd.read_csv(p_diablo_path, sep=\"\\t\")\n",
    "t_diablo = pd.read_csv(t_diablo_path, sep=\"\\t\")\n",
    "\n",
    "# load the diablo correlations\n",
    "diablo = pd.read_csv(diablo_path, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for technical reasons, we shortened names\n",
    "# we saved the mappings before running the pipeline, now we remap them\n",
    "l_fmap = pd.read_csv(l_fmap_path, sep=\"\\t\")[[\"short\", \"long\"]].set_index(\"short\")\n",
    "md_fmap = pd.read_csv(md_fmap_path, sep=\"\\t\")[[\"short\", \"long\"]].set_index(\"short\")\n",
    "mt_fmap = pd.read_csv(mt_fmap_path, sep=\"\\t\")[[\"short\", \"long\"]].set_index(\"short\")\n",
    "p_fmap = pd.read_csv(p_fmap_path, sep=\"\\t\")[[\"short\", \"long\"]].set_index(\"short\")\n",
    "t_fmap = pd.read_csv(t_fmap_path, sep=\"\\t\")[[\"short\", \"long\"]].set_index(\"short\")\n",
    "fmap = pd.concat([l_fmap, md_fmap, mt_fmap, p_fmap, t_fmap])\n",
    "fmap.index.name = None\n",
    "\n",
    "# R quietly converts non-alphanumeric characters in feature names to \".\"\n",
    "# we need to match these changes in order to remap\n",
    "fmap.index = [re.sub('[^0-9a-zA-Z_]', '.', x) for x in fmap.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(14, 2), (16, 2), (13, 2), (37, 2)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def remap_identifiers(data, keys):\n",
    "    data = keys.merge(data, left_index=True, right_index=True, how=\"right\")\n",
    "    data.set_index(\"long\", inplace=True)\n",
    "    data.sort_values(\"importance\", inplace=True, ascending=False)\n",
    "    data.index.name = None\n",
    "    return data[[\"GroupContrib\", \"importance\"]]\n",
    "\n",
    "singleomics = [remap_identifiers(x, fmap) for x in [l_splsda, m_splsda, p_splsda, t_splsda]]\n",
    "multiomics = [remap_identifiers(x, fmap) for x in [l_diablo, m_diablo, p_diablo, t_diablo]]\n",
    "[x.shape for x in multiomics]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigate how many of our single omics features match the original data (219 informative features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 0, 0, 0]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compare_data(old, new):\n",
    "    return old.merge(new, left_index=True, right_index=True, how=\"left\")\n",
    "\n",
    "# load in the most informative features identified by the original analysis\n",
    "features = pd.read_csv(features_path, sep=\"\\t\").set_index(\"standardized_name\")\n",
    "features.index.name = None\n",
    "features.index = [x.strip() for x in features.index.tolist()]\n",
    "\n",
    "# investigate how many single omics features match the authors data\n",
    "# for lipidomics, metabolomics, proteomics, transcriptomics respectively\n",
    "[compare_data(features, x).importance.nunique() for x in singleomics]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigate how many of our multi omics features match the original data (219 informative features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 0, 0, 0]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# investigate how many multi omics features match the authors data\n",
    "# for lipidomics, metabolomics, proteomics, transcriptomics respectively\n",
    "[compare_data(features, x).importance.nunique() for x in multiomics]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_1</th>\n",
       "      <th>Feature_2</th>\n",
       "      <th>Correlation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5589</th>\n",
       "      <td>PGD</td>\n",
       "      <td>PGD</td>\n",
       "      <td>0.922219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1701</th>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>0.922152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5541</th>\n",
       "      <td>PGD</td>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>0.919925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1749</th>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>PGD</td>\n",
       "      <td>0.919925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4617</th>\n",
       "      <td>H6PD</td>\n",
       "      <td>H6PD</td>\n",
       "      <td>0.918531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6389</th>\n",
       "      <td>ZNF527</td>\n",
       "      <td>PGD</td>\n",
       "      <td>-0.893038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4639</th>\n",
       "      <td>H6PD</td>\n",
       "      <td>ZNF527</td>\n",
       "      <td>-0.893972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6377</th>\n",
       "      <td>ZNF527</td>\n",
       "      <td>H6PD</td>\n",
       "      <td>-0.893972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6341</th>\n",
       "      <td>ZNF527</td>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>-0.897951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1759</th>\n",
       "      <td>unknown RT 16.698449</td>\n",
       "      <td>ZNF527</td>\n",
       "      <td>-0.897951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6400 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Feature_1             Feature_2  Correlation\n",
       "5589                   PGD                   PGD     0.922219\n",
       "1701  unknown RT 16.698449  unknown RT 16.698449     0.922152\n",
       "5541                   PGD  unknown RT 16.698449     0.919925\n",
       "1749  unknown RT 16.698449                   PGD     0.919925\n",
       "4617                  H6PD                  H6PD     0.918531\n",
       "...                    ...                   ...          ...\n",
       "6389                ZNF527                   PGD    -0.893038\n",
       "4639                  H6PD                ZNF527    -0.893972\n",
       "6377                ZNF527                  H6PD    -0.893972\n",
       "6341                ZNF527  unknown RT 16.698449    -0.897951\n",
       "1759  unknown RT 16.698449                ZNF527    -0.897951\n",
       "\n",
       "[6400 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reformat the correlation matrix to a more readable format\n",
    "diablo_flat = diablo.stack().reset_index()\n",
    "diablo_flat = diablo_flat.merge(fmap, left_on=\"level_0\", right_index=True, how=\"left\")\n",
    "diablo_flat.drop([\"level_0\"], axis=1, inplace=True)\n",
    "diablo_flat = diablo_flat.merge(fmap, left_on=\"level_1\", right_index=True, how=\"left\")\n",
    "diablo_flat.drop([\"level_1\"], axis=1, inplace=True)\n",
    "diablo_flat = diablo_flat[[\"long_x\", \"long_y\", 0]].sort_values(0, ascending=False)\n",
    "diablo_flat.columns = [\"Feature_1\", \"Feature_2\", \"Correlation\"]\n",
    "diablo_flat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare our highly correlating multi-omics feature values to original data (219 informative features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many of our highly correlating values match the authors?\n",
    "feature_list = diablo_flat[\"Feature_1\"].unique()\n",
    "\n",
    "feat_1 = features.merge(diablo_flat, left_index=True, right_on=\"Feature_1\", how=\"left\")\n",
    "feat_1 = feat_1[feat_1.Correlation.notnull()]\n",
    "\n",
    "feat_2 = features.merge(diablo_flat, left_index=True, right_on=\"Feature_2\", how=\"left\")\n",
    "feat_2 = feat_2[feat_2.Correlation.notnull()]\n",
    "\n",
    "feat_intersect = pd.concat([feat_1, feat_2]).reset_index().drop(\"index\", axis=1)\n",
    "feat_intersect.biomolecule_id.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find highly +ve/-ve correlations for drug target search (downstream analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H6PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C1orf226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Quinolinic acid, bis(trimethylsilyl) ester RT1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>LOC100130520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>X6R8F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>P80188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>P80188-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>CYFIP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0                                                 PGD\n",
       "1                                                H6PD\n",
       "2                                            C1orf226\n",
       "3                                              ZNF527\n",
       "4   Quinolinic acid, bis(trimethylsilyl) ester RT1...\n",
       "..                                                ...\n",
       "65                                       LOC100130520\n",
       "66                                             X6R8F3\n",
       "67                                             P80188\n",
       "68                                           P80188-2\n",
       "69                                             CYFIP2\n",
       "\n",
       "[70 rows x 1 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find highly +ve/-ve correlations for drug target search\n",
    "high_corr = diablo_flat[(diablo_flat.Correlation >= 0.5) | (diablo_flat.Correlation <= -0.5)]\n",
    "high_features = pd.DataFrame(high_corr[[\"Feature_1\", \"Feature_2\"]].stack().unique())\n",
    "outfile_path = \"/\".join([results_dir, \"high_correlated_features.txt\"])\n",
    "\n",
    "# map unknowns to putative unknowns\n",
    "putative_unknowns = pd.read_csv(unknowns_path, sep=\"\\t\", skiprows=1)\n",
    "putative_unknowns = putative_unknowns[[\"standardized_name\", \"Name\"]]\n",
    "high_features = high_features.merge(putative_unknowns, left_on=0, right_on=\"standardized_name\", how=\"left\")\n",
    "renamed = high_features[high_features.standardized_name.notnull()]\n",
    "for i in renamed.index:\n",
    "    high_features.loc[i] = renamed.loc[i][\"Name\"]\n",
    "high_features.drop([\"standardized_name\", \"Name\"], axis=1, inplace=True)\n",
    "high_features = high_features[~high_features[0].str.startswith(\"unknown\")]\n",
    "high_features = high_features[~high_features[0].str.startswith(\"Unknown\")]\n",
    "\n",
    "# some features are represented as a concatenation of multiple features\n",
    "high_features = high_features[0].tolist()\n",
    "high_features = [x.split(\";\") for x in high_features]\n",
    "high_features = [x for y in high_features for x in y]\n",
    "high_features = pd.DataFrame(high_features)\n",
    "high_features.to_csv(outfile_path, sep=\"\\t\", index=None, header=None)\n",
    "high_features.shape\n",
    "high_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare our highly correlating multi-omics feature values to authors (features ranked by importance in their tree-based classifier)\n",
    "\n",
    "We show that our highly correlated multi-omics features overlap with highly-scoring features identified by a tree-based classifier in the original analysis. First, we filtered this list by removing all features which were assigned an importance score of 0. Next, we matched the features we discovered through our pipeline to these features found by the original analysis. We find that a subset of 24 features fall above the median score of importance among important features identified in the original analysis. In particular, our pipeline recovers quinolinic acid, which was an important metabolite highlighted in the original study () due to its role in immune function (Heyes et al., 1992; Sofia et al., 2018) and COVID-19 severity (Migaud et al., 2020). Therefore, we show that our pipeline independently recapitulates some of the higher scoring features from an unrelated method which was customised to the experimental context. | As an additional layer of validation, we show that our feature list also overlaps with those yielded by an independent COVID-19 model (Migaud et al., 2020). |\n",
    "\n",
    "Our list of features is comparatively small compared to the full range of feature scores generated by the original study. However, we note that our approach works by selecting subsets of features with cross-validation. When contrasted against the original data, a subset of the most informative features were selected. Furthermore, our generic pipeline is shown to partially recapitulate results obtained from an algorithm which was specifically tailored to thee study. | Therefore, we demonstrate that our aim of providing an agnostic, flexible pipeline for analysing multi-omics data is met. |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAEICAYAAABxiqLiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAfn0lEQVR4nO3de5gcVZ3/8feHQAJeSIKEEQNC0LBu4iqLI+A+oMOCEhCMNzTCykXdGAVX96c+xBXvsD9UvIGYGJUfomRjvIAhooi6s+JiFFiQNUAwgJhAQC4hARIICd/fH+cM1DQ90zUn0zPJ5PN6nn6muupce6r7W3VOVbciAjMzs4HabrgbYGZmWycHEDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiIyKASLpA0hnD3Y6hIqlD0q8lPSTpC8Pdnq2BpJMkXTDc7RjpJHVLetcw1LuTpEslrZH0/aGuf1u1VQQQSX+WtF7Sw5JWS/qJpD2Hu13DaCZwH7BzRHxwcwraUoOvpDGSviXpjhwor5N0ZB9pPyEpJB2+mXWeJOl/Ja2TdLekOZLGbU6ZI5mkT0r67nC3I3sz0AE8JyKO3ZyCtrB+9SLpLZKuyvtod5Pt8yQtk/SEpJMats3Nn6E9j8ckPdSQZoakmyQ9IulWSYf0156tIoBkx0TEs4DdgXuAc4e5PUNC0qgmq/cCbowt4C5QSdu3qejtgRXAq4CxwMeAhZL2bqj/BaQPj1WbU5mkDwKfBT6c6zuI9DpfIWl0H3na1fdtXsFruxdwS0RsbEd7BqLN+8UDwJeBs/rY/gfgvcD/NG6IiFkR8ayeB/AfwJNna5JeTXoPnAw8G3glcFu/rYmILf4B/Bk4vPL8KNLO0vP8AuCMvHwS8JuG/AG8MC+PAc4G/kIKRHOBnQa5vWPyP/mu/PgyMKZm+y4A5gCXAY9U+13Z/jiwAXgYOJx0IDAbuBW4H1gI7FLJ833gbmAN8Gtgal4/s6GsSxvb0+T17QJWAqflMr/TX/3AjsB38/oHgauBjsLX9QbgTQ3rfpr3h177SJO8JwEX9LFt59z/tzSsfxbwV+Ad+fkngR/k/qwF3gUcAPw2920V8FVgdMP/dhbwJ2A1cB6gvG0U8AXS2eTtwKk5/fZ5+1jgW7ncO4EzgFGDvK++Frgu92cF8MnKti5gZbP3IjAt7zeP59fuD3l7N/AZ4L+Bh4CfA7tW8r8OWJpfr27gbxvKPi3/nx8jHUSclvv+ELAMOKxJHz7V0JZ35vXvAG7Kr/vlwF6VPF/J/V0LXAscktf31a9e+1feF76bl/fO/7d3kj5Xft1f/YCAL5H2rTW5vy8e4P/tXUB3P9t/A5zUz/Zn5tf0VZV1V/W8dnUfW9MZCACSngG8FVhSWMRngX2B/YAXAhOBj/dR18GSHuzncXAfdXyUdAS7H/BS0ofM6QNo43HAmaSjgN9UN0TEScBFwOciHUn8AvgX4PWko/Xn8dQHVY+fApOB3UhHJhflsuY1lHVMzfY9F9iFdNQ3s0X9J5I+CPcEnkP6MF0PIGm2pMV1KpTUQfq/La2sOxbYEBGX1Wx3X/6BFOh+VF0ZEQ+TXrtXV1ZPJwWRcaTXbhPwr8CuwCuAw0hHgFVHAy8n7QtvAY7I6/8ZOJK0n+xPeg2rvg1sJO2nfw+8hvTB8TSSjmuxrz6/j74/ApyQ+/Na4D2SGtvxNBHxM+Dfge/lfeellc3HkY5idwNGAx/KbdyXdNT7AWAC6SDp0oYzvLfldowDXkAKqi+PiGeTXrc/N2nLJxra8q3ch38D3pjrujLX3eNq0uu+CzAf+L6kHVv0q5VXAX8LHNGi/teQju73zf18K+kAq+f/eMMA6iz1JuBe0gFlz0hHJzBB0nJJKyV9VdJO/ZYymEcz7XqQdpqHSUctG0lH9X9X2X4BNc5ASJH/EeAFlW2vAG4f5PbeChxVeX4E8OdW7av05cIW5T/Z3/z8JipHZqRhvsfJR7INecfl+sY2K6uxPU1e3y7SEdqOdeonHYVdBbxkM17PHYBfAF+vrHsW6ah+UmUfKT0D+Sfg7j62nQVckZc/ST667KeeDwAXN7yWB1eeLwRm5+VfAe+ubDs8p9+eNJ7/GJWzY9KH638O5r7apP1fBr5U+V83PQOpvB7fbdjeDZxeef5e4Gd5+WPAwsq27UhnF12Vst9R2f5C0lH64cAOLdrdqy2kwP/OhrrWUTkLaci/GnhpP/3qtX/R/Axknzr1A/8I3EI6yNyu8P+0uWcgv6T32ebzch+uIb1/dyWdRZ7ZXzu2pjOQ10fEONLw0KnAf0l67gDLmAA8A7i258gM+FleP5ieB9xReX5HXlfXigHWtxdwcaVPN5GOjDskjZJ0Vp4QW8tTR3C7DrCOqnsj4tE69ZOGuC4HFki6S9LnJO1QtyJJ2+UyNpD+7z0+BXwnIm7fjH70uA/YtY+x693z9h69/jeS9pW0OE+6ryUdvTa+tndXlteRgh+kfaJaXnV5L1LgXFV5Xb9OOqofNJIOlPSfku6VtIZ0hrg5+wb0398n3xcR8QSpzxMr6VdUti8nBeRPAn+VtEBS3ffRXsBXKq/dA6QDyImQ5rzyZPGavH0sm9/vxv9f0/oj4lekoc7zgHvyxPfOm1l3bfkCpFcBF1ZWr89/z42IVRFxH/BF0vBwn7amAAJARGyKiB+RPqCaDSE9QgoSADQEmftIL9TUiBiXH2MjTSg9jaRDGq5aaHz0dYXCXaQdqMfz87pW7Xuym32U25cVwJGVPo2LiB0j4k7ScMJ00lHcWNLREqSdua+61lXbSBqy6q99fdYfEY9HxKciYgppqOho0pBJS5JEmgPoIM19PF7ZfBjwL/mD+27SENlCSafVKbvBb0lH+29sqP+ZpCGmX1ZWN/Z9DnAzMDkidiYNW4h6VgF7VJ5Xryxckdu0a+U13TkipjYrSNLxLfbVvoaw5gOLgD0jYixpTrCn/Y376ih6H2wNdD/t9b7I/989SWchTcuMiPkRcXDOF6Qh6DpWkM7uqvvkThFxVX7fnkYaThyfD0zX0P97otdrwdPfE435+qw/9+uciHgZMJU0lPXhmv0aDCcAV0XEkxPkEbGaNLc5oP/pVhdAlEwHxpOOdBv9AZgqaT9JO5KOXoAnj3i+AXxJ0m65vImSjmhSDhFxZVSuWmjyuLKPZv4HcLqkCZJ2Jc2x9FwW2Gf7NsNc4ExJe+U+TcivEaR5lMdIY6zPIB0hV90D7NOw7nrguHz2Mo10tFJUv6RDJf1d/vBZSxra2lSzX3NIY8rHRMT6hm2HAS8mjWPvR/pweje9535qiYg1pDOacyVNk7SD0tVe3ye9qb7TT/Znk/r1sKQXAe8ZQNULgffnfXAc6UOtp02rSBPQX5C0s6TtJL1AUtP/RURc1GJf/Us/7X8gIh6VdADpgKPHLcCOkl6bzxpPJ40A9LgH2DufJdbt72slHZbL+yBp37yqWWJJfyPpHyWNAR4lHfzV3XfmAh+RNDWXNVZpzgxSnzeS5gC2l/Rx0oUU/fXremBG3jc6SVf+FdUv6eX5zG8HUmB6tG6/8ntyR9Iw53aSdqye0UsanbcL2CFvb/z/nEAalm70/4D3SdpN0njS2V+/c5RbUwC5VNLDpDfrmcCJEbG0MVFE3AJ8mjRm/icaJqFJb9LlwJI85PAL4G8Gua1nkMYSbwD+lzRxfUbN9pX4Cuko8udK13UvAQ7M2y4kDRvcCdzI0y8++BYwJZ9qX5LXvR84hjTndDxwCf3rr/7nkiad15IC/n+Rg6mkf5P002YF5mD0blJwuLtyJH08QETcHxF39zxIb8DVkSa+BywiPkc6ezg7t/V3pKPIwyLisX6yfoj0ofsQ6eDkewOo9hukIHED6Uqoy0gfbD0fJieQJqFvJI3R/4A0pDaY3gt8Ov/fPk76kAeeDKzvBb5J2n8eIQXUHj2XgN4v6WmXjTaKiGWk+aZzSaMBx5AODjb0kWUMaQ7qPtKw2G6k/1FLEXEx6WxlQX6f/5F0NglpSPWnpAB5B+kDvDr81KxfHyNN6q8mHWzM34z6dyb971fn+u8n7Xc9Z5JP+1yreDspkM4BDsnL36hs/3le9w/AvLz8yp6Nkl5BOuttdrPlZ0gXF9xCeq9eR/qs7VPP5YRmI5rSTVVdka5i2yIp3Sg5NyL2apnYbAuwNZ2BmI0oSl+/cZSk7SVNBD4BXDzc7TKrq1YAyePCy5SuD57dZLsknZO33yBp/1Z5JX1e0s05/cV5DBhJeyt9bcn1+TF3MDpq27zraT0UN9REGg5ZTRouuIk+7kky2xK1HMLKk5+3kG6mWkkaI3tbRNxYSXMU8D7SJV8HAl+JiAP7yyvpNcCvImKjpM8CRMRpefJycUS8eFB7amZmg6rOGcgBwPKIuC1Pdi0gXRZaNZ1081tExBJgnKTd+8sbET+Pp763Zgm9L2c0M7MtXJ0v/ZpI7ysUVvLUFTb9pZlYMy+ku5WrV69MktTz/Tyn93O5LAC77rpr7L333v0lMRs2a9asYezYscPdDLOnufbaa++LiOIbqesEkGY3RTWOe/WVpmVeSR8lXbp4UV61Cnh+RNwv6WXAJZKmRsTahnwzSd/DREdHB2effXbLjpgNBwcQ21Ideuihd7RO1bc6AWQlve+Q3YOn7qpulWZ0f3klnUi6M/mwyJMx+Zr7x/LytZJuJd2peU21wkhfBDgPoLOzM7q6ump0xWzoLV68GO+fNhLVmQO5GpgsaZLSt2bOIN00VrUIOCFfjXUQsCbfSdtnXqU7nE8DXhcR63oKyncxj8rL+5C+Rbb/76Q3M7Mh1/IMJF8ldSrp7s1RwPkRsVTSrLx9LukO2qNId3ivI32Vc595c9FfJd1peoUkgCURMYt01+SnJfXckTsrIh4YrA6bmdngqPXLWZF+b+GyhnVzK8sBnFI3b17/wj7S/xD4YZ12mZnZ8PGd6GZmVsQBxMzMijiAmJlZEQcQMzMrUmsS3cyy+XV/bPApR0OLX49o4jj/zIJt+RxAzAai5IN9vhwQbETyEJaZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFfB+I2QBo4PcRAgHHDzCHbxuxrYADiNkAlHywSw4INjJ5CMvMzIo4gJiZWREHEDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWREHEDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWZFaAUTSNEnLJC2XNLvJdkk6J2+/QdL+rfJK+rykm3P6iyWNq2z7SE6/TNIRm9tJMzMbfC0DiKRRwHnAkcAU4G2SpjQkOxKYnB8zgTk18l4BvDgiXgLcAnwk55kCzACmAtOAr+VyzMxsC1LnDOQAYHlE3BYRG4AFwPSGNNOBCyNZAoyTtHt/eSPi5xGxMedfAuxRKWtBRDwWEbcDy3M5Zma2Bdm+RpqJwIrK85XAgTXSTKyZF+AdwPcqZS1pUlYvkmaSznbo6Oigu7u7RTfMhkuX908bkeoEEDVZFzXTtMwr6aPARuCiAdRHRMwD5gF0dnZGV1dXk2xmWwbvnzYS1QkgK4E9K8/3AO6qmWZ0f3klnQgcDRwWET1Bok59ZmY2zOrMgVwNTJY0SdJo0gT3ooY0i4AT8tVYBwFrImJVf3klTQNOA14XEesaypohaYykSaSJ+d9vRh/NzKwNWp6BRMRGSacClwOjgPMjYqmkWXn7XOAy4CjShPc64OT+8uaivwqMAa6QBLAkImblshcCN5KGtk6JiE2D1mMzMxsUemrkaOvV2dkZ11xzzXA3w6wpCUbA28xGIEnXRkRnaX7fiW5mZkUcQMzMrIgDiJmZFXEAMTOzIg4gZmZWpM6NhGYj1i67wOrV7a9Hzb5fYRCNHw8PPNDeOswaOYDYNm316iG4xHa+4Lj2VtLuAGXWjIewzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWREHEDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWREHEDMzK+IAYmZmRfyDUrbNa/+PMQUc3+46zIaeA4ht8/yLhGZlPIRlZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMrUiuASJomaZmk5ZJmN9kuSefk7TdI2r9VXknHSloq6QlJnZX1e0taL+n6/Ji7uZ00M7PB1/JOdEmjgPOAVwMrgaslLYqIGyvJjgQm58eBwBzgwBZ5/wi8Efh6k2pvjYj9yrtlZmbtVucM5ABgeUTcFhEbgAXA9IY004ELI1kCjJO0e395I+KmiFg2aD0xM7MhVee7sCYCKyrPV5LOMlqlmVgzbzOTJF0HrAVOj4grGxNImgnMBOjo6KC7u7tGsWaNutq+73TBEOyf7e+HWaM6AaTZ17Q1fjNcX2nq5G20Cnh+RNwv6WXAJZKmRsTaXoVEzAPmAXR2dkZXV1eLYs2aa/u+M38I6mBo6jCrqjOEtRLYs/J8D+Cummnq5O0lIh6LiPvz8rXArcC+NdppZmZDqE4AuRqYLGmSpNHADGBRQ5pFwAn5aqyDgDURsapm3l4kTciT70jahzQxf9uAemVmZm3XcggrIjZKOhW4HBgFnB8RSyXNytvnApcBRwHLgXXAyf3lBZD0BuBcYALwE0nXR8QRwCuBT0vaCGwCZkXEA4PZaTMz23yKtv+aTvt1dnbGNddcM9zNsK2QNHJ+UGoEvJVtiEm6NiI6W6dszneim5lZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVqTOd2GZjWhq9o1tgyrg+PbWMH58e8s3a8YBxLZpQ3HznW/ys5HKQ1hmZlbEAcTMzIo4gJiZWREHEDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWREHEDMzK+IvU7Rt2w92gQ2r21xJwPw2f+Xv6PHw5gfaW4dZAwcQ27ZtWA3Htfmrco+n/XW0O0CZNeEhLDMzK+IAYmZmRRxAzMysiAOImZkVcQAxM7MiDiBmZlbEAcTMzIo4gJiZWREHEDMzK+IAYmZmRWoFEEnTJC2TtFzS7CbbJemcvP0GSfu3yivpWElLJT0hqbOhvI/k9MskHbE5HTQzs/ZoGUAkjQLOA44EpgBvkzSlIdmRwOT8mAnMqZH3j8AbgV831DcFmAFMBaYBX8vlmJnZFqTOGcgBwPKIuC0iNgALgOkNaaYDF0ayBBgnaff+8kbETRGxrEl904EFEfFYRNwOLM/lmJnZFqTOt/FOBFZUnq8EDqyRZmLNvM3qW9KkrF4kzSSd7dDR0UF3d3eLYs2erguGYN/pansdXQxFP8x6qxNAmn1PdON3U/eVpk7ekvqIiHnAPIDOzs7o6upqUaxZE/NhKPadttcxRP0wq6oTQFYCe1ae7wHcVTPN6Bp5S+ozM7NhVmcO5GpgsqRJkkaTJrgXNaRZBJyQr8Y6CFgTEatq5m20CJghaYykSaSJ+d8PoE9mZjYEWp6BRMRGSacClwOjgPMjYqmkWXn7XOAy4CjShPc64OT+8gJIegNwLjAB+Imk6yPiiFz2QuBGYCNwSkRsGtRem5nZZlNEm39qcwh0dnbGNddcM9zNsK3RfLX952YlaPvbbAj6YSOPpGsjorN1yuZ8J7qZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZke2HuwFmw26+2lp8XATMb2sVZsPCAcTsuGhr8RJEe6toexA0a8ZDWGZmVsQBxMzMijiAmJlZEQcQMzMr4kl0swFQ4Vz1QPO1fdLdbBD4DMRsACIG/rj00sUDzmO2NXAAMTOzIg4gZmZWxAHEzMyKOICYmVmRWgFE0jRJyyQtlzS7yXZJOidvv0HS/q3yStpF0hWS/pT/js/r95a0XtL1+TF3MDpqZmaDq2UAkTQKOA84EpgCvE3SlIZkRwKT82MmMKdG3tnALyNiMvDL/LzHrRGxX37MKu2cmZm1T50zkAOA5RFxW0RsABYA0xvSTAcujGQJME7S7i3yTge+nZe/Dbx+M/tiZmZDqM6NhBOBFZXnK4EDa6SZ2CJvR0SsAoiIVZJ2q6SbJOk6YC1wekRc2dgoSTNJZzt0dHTQ3d1doytmvXVB2/edTZs2tb2OLtrfD7NGdQJIs3toG2916itNnbyNVgHPj4j7Jb0MuETS1IhY26uQiHnAPIDOzs7o6upqUaxZE/Oh3fvO4sWL217HUPTDrFGdIayVwJ6V53sAd9VM01/ee/IwF/nvXwEi4rGIuD8vXwvcCuxbpzNmZjZ06gSQq4HJkiZJGg3MABY1pFkEnJCvxjoIWJOHp/rLuwg4MS+fCPwYQNKEPPmOpH1IE/O3FffQzMzaouUQVkRslHQqcDkwCjg/IpZKmpW3zwUuA44ClgPrgJP7y5uLPgtYKOmdwF+AY/P6VwKflrQR2ATMiogHBqW3ZmY2aGp9G29EXEYKEtV1cyvLAZxSN29efz9wWJP1PwR+WKddZmY2fHwnupmZFXEAMTOzIg4gZmZWxAHEzMyKOICYmVkRBxAzMytS6zJesxFtfrNv3Bk8RwPMb2sVMHp8myswezoHENu2Hdfqq9kGwXwNTT1mQ8xDWGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVsQBxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKyIA4iZmRVxADEzsyIOIGZmVqRWAJE0TdIyScslzW6yXZLOydtvkLR/q7ySdpF0haQ/5b/jK9s+ktMvk3TE5nbSzMwG3/atEkgaBZwHvBpYCVwtaVFE3FhJdiQwOT8OBOYAB7bIOxv4ZUSclQPLbOA0SVOAGcBU4HnALyTtGxGbBqfLZpthvoYm33FRVo/ZEGoZQIADgOURcRuApAXAdKAaQKYDF0ZEAEskjZO0O7B3P3mnA105/7eBbuC0vH5BRDwG3C5peW7Db8u7aTZICj7YFy9ezNFHH92GxpgNrzoBZCKwovJ8Jekso1WaiS3ydkTEKoCIWCVpt0pZS5qU1YukmcBMgI6ODrq7u2t0xWzobdq0yfunjUh1Akizc+/Gw7C+0tTJW1IfETEPmAfQ2dkZXV1dLYo1Gx6LFy/G+6eNRHUm0VcCe1ae7wHcVTNNf3nvycNc5L9/HUB9ZmY2zOoEkKuByZImSRpNmuBe1JBmEXBCvhrrIGBNHp7qL+8i4MS8fCLw48r6GZLGSJpEmpj/fWH/zMysTVoOYUXERkmnApcDo4DzI2KppFl5+1zgMuAoYDmwDji5v7y56LOAhZLeCfwFODbnWSppIWmifSNwiq/AMjPb8tSZAyEiLiMFieq6uZXlAE6pmzevvx84rI88ZwJn1mmbmZkND9+JbmZmRRxAzMysiNLo09ZN0r3AHcPdDrM+TAJuH+5GmDWxV0RMKM08IgKI2ZZM0iMR8czhbofZYPMQlpmZFXEAMTOzIg4gZu33o+FugFk7eA7EzMyK+AzEzMyKOICYmVkRBxDb4kkKSesrj4MLyvi/ko5pU/v+XtLKdpTdT51vlfTxoazTrJHnQGyLJykiovC3ZJ8sYzlwSUR8aAB5xuRfxtyiSBpD+qnoAyLiJcPdHtt2+QzEtkqSjpf0oKRHJN0n6aV5/YV53XpJd0p6Tv71yn2A9+f1h+a8b8959pW0MS9/U9IKSXcDqyRNkHRLLnOdpKd9yaekgyU9Wsm/StLdkh6XtFDSj3PehyTtk9M9KOk6SWslPSrppLx+n5x/fU7/pry+W9JNku4DlgEnAS/O6b4s6aRc1rr8d1qlPXdKulfSBkm/q7T7ozn9ekkP5HUt+2v2pIjww48t+kH6Rcr1+XEXsBOwFnhR3v5l4Ja8/MJKviuB7+fl5cDZlW0PAm/Py/sCG/PyN0k/I7BPfv7fwNfy8l7ABmBCQ/sOBh6t5N8A7A68KLf9orztf4CLK/XfnJdPreT/A/CrvPx/gPV5uRt4BBhfqeeGShsmAmPy8oeBOyvpHif9SNvY3LcDcts2AofkdLX764cfPQ+fgdhWISJ2yo/nAUcAzwauk7QemAXskpMeJWlNPiM4CJhSUN3tEXFbXt4PeEeu52bSWfsBLfLfFhGrIuJmYBPwxbz+emDvSrp5uW9fBbaXtBfpB9Q+ldd/EdhBUs8vdP4uIlb3UedE4Pbc788Au1a23RoRKyJiDbAaeBlwHLAiIq7MdW1Of20b5QBiW6PtgIcrQWXHiOj5wPwC8K6I2BH4NjCmjzI2kX7kDGDnhm3rG56/vlLX9hHxkxbt29Dw/KH894lKnZDOTqqeAJrN9TyR/z7ST53zgatyvw9vqKfanidIr8l2TervMdD+2jbKAcS2RpcDYyT9M4CknSS9Lm8bBSyVtBPwhkqeR4Dxlef3Aq/Jyx/tp67rgc9JUq5rxiC0v8e7cpnvIQ2hrSDNb3wsr/8AsCEi7mySdzXwjMrzZ/DUN/6eUaPu7wLPl3RIrmufvL6d/bURxgHEtjoR8QjwduDzeahlNfDmvHkBaR5hFVC9tHYO8E95YvhQ4F+BN0haS+/hnkZHk365c30eHvrKIHZlTa7/S6RhOIA3AlNzvz5D6mczc4DdeybRgU+QLhJYS433dR5e+xRwea7r6rypnf21EcaX8ZoNA0kPAu+LiO8Md1vMSvkMxMzMivgMxMzMivgMxMzMijiAmJlZEQcQMzMr4gBiZmZFHEDMzKzI/wdC7eDvDAjrJAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Features with positive importance score in original analysis (of all features)=8.090258668134286 %\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(24, 2)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# where do our features fall in the spectrum of importance shown by an independent tree-based classifier?\n",
    "feature_importance_all = pd.read_csv(feature_importance_all_path, sep=\"\\t\")\n",
    "total_features = feature_importance_all[\"feature importance\"].count()\n",
    "\n",
    "# remove all features with null importance score\n",
    "feature_importance_all = feature_importance_all[feature_importance_all[\"feature importance\"] > 0]\n",
    "\n",
    "# some features are composite (eg different isoforms of same protein)\n",
    "feature_importance_all = pd.DataFrame(pd.concat([pd.Series(row['feature importance'], \n",
    "                                                           row['biomolecule'].split(';')) \n",
    "                                                for _, row in feature_importance_all.iterrows()]))\n",
    "\n",
    "# some features we detect overlap with the authors, show distribution\n",
    "fig, ax = plt.subplots()\n",
    "feature_importance_all.boxplot(color=\"orange\", ax=ax, showfliers=False)\n",
    "intersecting_features = high_features.merge(feature_importance_all, left_on=0, right_index=True)\n",
    "intersecting_features.drop(\"0_x\", axis=1, inplace=True)\n",
    "intersecting_features.columns = [\"index\", \"Feature importance\"]\n",
    "intersecting_features.boxplot(color=\"blue\", showbox=True, ax=ax, showfliers=False)\n",
    "axis_title = \" \".join([\"Blue = our features:\", \n",
    "                       str(intersecting_features.count()[0]),\n",
    "                       \"| Orange = authors features:\",\n",
    "                       str(feature_importance_all.count()[0]),\n",
    "                      ])\n",
    "ax.title.set_text(axis_title)\n",
    "plt.show()\n",
    "\n",
    "percent_important = str((feature_importance_all.count()[0] / total_features) * 100)\n",
    "print(\"\".join([\"Features with positive importance score in original analysis (of all features)=\", \n",
    "               percent_important,\n",
    "              \" %\"]))\n",
    "\n",
    "# features that dont have positive importance scores are also interesting\n",
    "nonintersecting_features = high_features.merge(feature_importance_all, left_on=0, right_index=True, how=\"outer\")\n",
    "nonintersecting_features = nonintersecting_features[nonintersecting_features[\"0_y\"].isna()][\"key_0\"]\n",
    "nonintersecting_features = nonintersecting_features.reset_index().drop(\"index\", axis=1)\n",
    "intersecting_features.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "The files in the results directory were a concatenation of all features obtained from the analyses. These are available in gitlab in the results directory. To reproduce these files, after running our pipeline navigate to your results directory and run the following commands.\n",
    "\n",
    "## Authors data\n",
    "\n",
    "- [Supplementary table 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7388490/bin/90133-2020.07.17.20156513-2.xlsx) - enriched pathways in each omics\n",
    "- [Supplementary table 2](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7388490/bin/90133-2020.07.17.20156513-3.xlsx) - unknown features mapped to putative features\n",
    "- [Supplementary table 3](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7388490/bin/90133-2020.07.17.20156513-4.xlsx) - 219 informative features\n",
    "- [Supplementary table 6](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7388490/bin/90133-2020.07.17.20156513-7.xlsx) - feature importances identified by a tree-based classifier\n",
    "\n",
    "Files were exported from the excel spreadsheets as tab separated lists of values. Unrecognised byte characters and windows line endings were removed. The clean versions of the files in plain text format are available in the gitlab repository.\n",
    "\n",
    "\n",
    "## Concatenating files to examine features\n",
    "\n",
    "The pipeline outputs features per individual component. We concatenated all these to get the full feature lists returned by `sPLSDA` and `DIABLO`. These files are available in the git repository under `results/MSV000085703/reformatted_output/`\n",
    "\n",
    "```\n",
    "#!/bin/bash\n",
    "head -n1 lipidome_1_sPLSDA_max.txt > lipidome_sPLSDA_max.txt\n",
    "for i in lipidome_[0-9]_sPLSDA_max*; do tail -n +2 $i >> lipidome_sPLSDA_max.txt; done\n",
    "head -n1 metabolome_1_sPLSDA_max.txt > metabolome_sPLSDA_max.txt\n",
    "for i in metabolome_[0-9]_sPLSDA_max*; do tail -n +2 $i >> metabolome_sPLSDA_max.txt; done\n",
    "head -n1 proteome_1_sPLSDA_max.txt > proteome_sPLSDA_max.txt\n",
    "for i in proteome_[0-9]_sPLSDA_max*; do tail -n +2 $i >> proteome_sPLSDA_max.txt; done\n",
    "head -n1 transcriptome_1_sPLSDA_max.txt > transcriptome_sPLSDA_max.txt\n",
    "for i in transcriptome_[0-9]_sPLSDA_max*; do tail -n +2 $i >> transcriptome_sPLSDA_max.txt; done\n",
    "\n",
    "head -n1 lipidome_1_DIABLO_var_keepx_max.txt > lipidome_DIABLO_max.txt\n",
    "for i in lipidome_[0-9]_DIABLO_var_keepx_max*txt; do tail -n +2 $i >> lipidome_DIABLO_max.txt; done\n",
    "head -n1 metabolome_1_DIABLO_var_keepx_max.txt > metabolome_DIABLO_max.txt\n",
    "for i in metabolome_[0-9]_DIABLO_var_keepx_max*txt; do tail -n +2 $i >> metabolome_DIABLO_max.txt; done\n",
    "head -n1 proteome_1_DIABLO_var_keepx_max.txt > proteome_DIABLO_max.txt\n",
    "for i in proteome_[0-9]_DIABLO_var_keepx_max*txt; do tail -n +2 $i >> proteome_DIABLO_max.txt; done\n",
    "head -n1 transcriptome_1_DIABLO_var_keepx_max.txt > transcriptome_DIABLO_max.txt\n",
    "for i in transcriptome_[0-9]_DIABLO_var_keepx_max*txt; do tail -n +2 $i >> transcriptome_DIABLO_max.txt; done\n",
    "\n",
    "# md5sum if on linux machine\n",
    "md5 lipidome_sPLSDA_max.txt metabolome_sPLSDA_max.txt proteome_sPLSDA_max.txt transcriptome_sPLSDA_max.txt \\\n",
    "  lipidome_DIABLO_max.txt metabolome_DIABLO_max.txt proteome_DIABLO_max.txt transcriptome_DIABLO_max.txt\n",
    "\n",
    "MD5 (lipidome_sPLSDA_max.txt) = 46ad9dde93559c35bf4026184761675b\n",
    "MD5 (metabolome_sPLSDA_max.txt) = 53001f373a9163a8d1dea70f72546a08\n",
    "MD5 (proteome_sPLSDA_max.txt) = b336b47e238de50d32e0b3a523e07fd9\n",
    "MD5 (transcriptome_sPLSDA_max.txt) = 9d704fd469f985c17ac1ce907ce3ec37\n",
    "MD5 (lipidome_DIABLO_max.txt) = aaba4c96a3fcc77079da1652ae2c5a2d\n",
    "MD5 (metabolome_DIABLO_max.txt) = cfec4e72698cc3e6b90a5452a6ee4c76\n",
    "MD5 (proteome_DIABLO_max.txt) = 0b7879b44dc349663e6b2aa6e4ce8732\n",
    "MD5 (transcriptome_DIABLO_max.txt) = 6c809bc489ffb8ff433de1b0d14bd2bb\n",
    "```\n",
    "\n",
    "## Passing results to a drug target search database\n",
    "\n",
    "```\n",
    "# put these directly into a drug target search database\n",
    "wget 'http://unmtid-shinyapps.net/download/drug.target.interaction.tsv.gz'\n",
    "\n",
    "# some rows have multiple molecules - also note that mostly proteins are in this database\n",
    "(head -n1 drug.target.interaction.tsv; grep -f ../../results/MSV000085703/reformatted_output/high_correlated_features.txt drug.target.interaction.tsv | grep -v MOUSE) | tr -d '\"' > ../../results/MSV000085703/reformatted_output/potential_drugs.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
